util/etcd/election.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+// MasterElection is an implementation of util.MasterElection based on etcd.
+type MasterElection struct {
+	instanceID string
+	treeID     int64
+	lockFile   string
+	client     *clientv3.Client
+	session    *concurrency.Session
+	election   *concurrency.Election
+}
+
+// Start commences election operation.
+func (eme *MasterElection) Start(ctx context.Context) error {
+	return nil
+}
+
+// WaitForMastership blocks until the current instance is master.
+func (eme *MasterElection) WaitForMastership(ctx context.Context) error {
+	return eme.election.Campaign(ctx, eme.instanceID)
+}
+
+// IsMaster returns whether the current instance is the master.
+func (eme *MasterElection) IsMaster(ctx context.Context) (bool, error) {
+	leader, err := eme.election.Leader(ctx)
+	if err != nil {
+		return false, err
+	}
+	return leader == eme.instanceID, nil
+}
+
+// Resign relinquishes mastership.
```

Comment:
The interface docs says this'll rejoin, but this comment is strangely silent on that bit :)